File: App.MasterDataEditor/Helpers/GitBlameTiming.cs
using System;
using System.Diagnostics;
using System.Text.Json;

namespace App.MasterDataEditor;

/// <summary>BLAME の各段階の計測値だけを、requestId 付きでログと WebView に通知する。</summary>
public sealed class GitBlameTiming
{
	private readonly string _requestId;
	private readonly string _filename;
	private readonly Action<object> _send;

	public GitBlameTiming(string requestId, string filename, Action<object> send)
	{
		_requestId = requestId;
		_filename = filename;
		_send = send;
	}

	public void Record(string stage, double durationMs, long? chars = null, int? entryCount = null, bool success = true)
	{
		var message = new
		{
			type = "git_blame_timing", requestId = _requestId, filename = _filename,
			source = "host", stage, durationMs, chars, entryCount, success,
		};
		Logger.Info("[BLAME timing] " + JsonSerializer.Serialize(message));
		_send(message);
	}

	public void RecordSince(string stage, long startedAt, long? chars = null, int? entryCount = null, bool success = true)
	{
		Record(stage, Stopwatch.GetElapsedTime(startedAt).TotalMilliseconds, chars, entryCount, success);
	}
}

File: WebView/src/app/git-blame-timing.ts
/** BLAME の計測値だけを記録する。レスポンス本体を計測ログに複製しない。 */
export interface GitBlameTiming {
    type: 'git_blame_timing';
    requestId: string;
    source: 'host' | 'webview';
    stage: string;
    durationMs: number;
    success: boolean;
    [key: string]: unknown;
}

export interface GitBlameTimingContext {
    requestId: string;
    filename: string;
}

let timingSink: ((timing: GitBlameTiming) => void) | undefined;
let listening = false;

export function configureGitBlameTiming(sink: (timing: GitBlameTiming) => void): void {
    timingSink = sink;
    if (listening) return;
    listening = true;
    window.chrome.webview.addEventListener('message', (event: MessageEvent) => {
        // 自前の小さな計測通知だけを読む。巨大なBLAMEレスポンスの解析を増やさない。
        if (typeof event.data !== 'string' || !event.data.startsWith('{"type":"git_blame_timing",')) return;
        const timing = JSON.parse(event.data) as GitBlameTiming;
        publishTiming(timing);
    });
}

function publishTiming(timing: GitBlameTiming): void {
    console.info(`[BLAME timing] ${JSON.stringify(timing)}`);
    timingSink?.(timing);
}

export function recordGitBlameTiming(requestId: string, stage: string, durationMs: number, metrics: Record<string, unknown> = {}): void {
    publishTiming({type: 'git_blame_timing', requestId, source: 'webview', stage, durationMs, success: true, ...metrics});
}

/** 既存の各受信リスナーが行っている解析を、そのまま個別に計測する。 */
export function parseWebViewMessageWithBlameTiming(json: string, consumer: string): any {
    const startedAt = performance.now();
    const message = JSON.parse(json);
    const durationMs = performance.now() - startedAt;
    if (message?.type === 'git_blame_response') {
        recordGitBlameTiming(message.requestId, 'response_json_parse', durationMs, {consumer, chars: json.length});
    }
    return message;
}
